### ISSUE TITLE
Proposal: optional Cargo features to gate ONNX op categories (measured 8-14% WASM size reduction on DFN3)

### ISSUE THREAD

**Issue by czoli1976** [sonos/tract#2158]
## Summary

Hi @kali — sharing a separate item that came out of our DFN3 validation sweep. This is a **proposal + question**, not a bug. No urgency.

While benchmarking the tract upgrade path for our WASM streaming-audio workload, we tracked binary-size growth across versions and tried two stripping approaches. The data is interesting and there's a small upstream improvement we'd love to help with if you'd find it useful.

## Size growth observed across versions

After `wasm-opt -Oz` post-processing, building libDF's `df_bg.wasm` against each tract version:

| tract | WASM size | Δ vs 0.21.4 | what (likely) drove the change |
|---|---:|---:|---|
| 0.21.4 (Mezon's current pin) | 8.28 MiB | — | baseline |
| 0.21.5 | 8.47 MiB | +2.3% | minor |
| 0.21.6 | 8.85 MiB | +6.9% | first WASM f32 4×4 kernel + multithread-mm code |
| 0.21.12 | 9.14 MiB | +10.4% | full MMM kit + reduce-op variants |
| 0.21.15 | 9.13 MiB | +10.3% | MMM kit consolidated |
| 0.22.0 | 9.22 MiB | +11.4% | additional SIMD variants |
| 0.22.1 (our ship target) | 9.22 MiB | +11.4% | flat vs 0.22.0 (2-file diff) |
| **0.23.0-dev.5** | **12.33 MiB** | **+49%** | safetensors + getrandom 0.4 + prost 0.11→0.14 + itertools 0.10→0.14 |

**On the 0.21.5 → 0.22.1 trajectory (~11%)**: this looks proportional to the very real kernel improvements you've shipped — we're getting ~25% faster real-audio inference for ~11% larger WASM, which is a fine trade-off in our book. No concern there.

**On the 0.23.0-dev.5 +49% jump**: the dependency chain (safetensors brings in getrandom 0.4, plus prost/itertools major bumps) accounts for most of it as far as we can tell. **Quick question — is that delta expected to persist into 0.23.0 stable, or is some of it a temporary side-effect of dev-cycle dep choices that may settle before stable?** Asking partly because we'll re-evaluate 0.23 once it stabilizes and once #2157 (the PulsedModel regression) is fixed, and a +49% size jump would be the main concern after PulsedModel.

## What we tried for stripping (and what worked)

**Background**: tract has no per-op or per-category gating today — every consumer's WASM carries every ONNX op type, even ones their model never touches. For specialized workloads (streaming audio, single-purpose vision pipelines, etc.) this is a meaningful overhead.

We tried two approaches against tract 0.22.1, both verified to produce SHA256-byte-identical output to the unstripped build on our test corpus.

### Approach A — strip unused op CATEGORIES

In `tract-onnx-0.22.1/src/ops/mod.rs::register_all_ops`, skip categories DFN3 doesn't use:
- `ml::*` (CategoryMapper, TreeEnsembleClassifier — sklearn-style)
- `fft::*` (DFT, STFT, MelWeightMatrix — DFN3 does its own FFT)
- `multinomial`, `random::*`
- `non_max_suppression` (vision-specific)
- `s2d::*`, `d2s::*` (DepthToSpace/SpaceToDepth)
- `cumsum::*`, `logic::*`
- `quant::*`
- top-level `Resize`

Kept: `array`, `cast`, `math`, `nn`, `rec`, plus `Constant`, `Einsum`, `Identity`.

**Result: 9.22 MiB → 8.46 MiB (−8.2%, −796 KB).** Output byte-identical.

### Approach B — strip individual unused OPS within kept categories

Within `array`, `cast`, `math`, `nn`, `rec`, also skip:
- `nn`: LSTM/RNN, all Pool variants, BatchNorm/InstanceNorm/LayerNorm, LRN, all Reduce*, ArgMin/ArgMax, advanced activations
- `array`: Tile, Split, Scatter*, Gather variants, NonZero, OneHot, TopK, Trilu, Compress, etc.
- `math`: trig + transcendentals, BitShift/Bitwise*, Comparison ops, MatMulInteger/QLinearMatMul, Pow/Mod/Rem, etc.
- `rec`: LSTM, RNN, Scan (DFN3 only uses GRU)

**Result: 9.22 MiB → 7.89 MiB (−14.4%, −1.36 MB).** Still byte-identical output.

That's actually **smaller than the 0.21.4 baseline (8.28 MiB)** while still getting all of 0.22.1's kernel speedups. We can ship the upgrade with no size regression and a 25% RTF win.

### Why dead-code elimination alone (wasm-opt) doesn't do this

We checked: `wasm-opt -Oz` is already maxed out in our build. The reason it can't strip these ops is `register_all_ops` does:
```rust
reg.insert("CategoryMapper", |...| { ... });
```
The function pointer in the HashMap looks reachable to the static analyzer → the function and its dependencies survive DCE → all of `tract_onnx::ops::ml::*` is kept. Only by *not registering* the op (compile-time choice) does it become genuinely unreachable.

## Proposal — graduated upstream paths

In rough order of effort vs benefit:

**A. Per-category Cargo features** (12 features)
- ~50-line diff: 12 entries in `Cargo.toml [features]`, 12 `#[cfg(feature = "...")]` lines in `register_all_ops`
- `default = ["all-onnx-ops"]` keeps current behavior identical for existing users
- Saves 5–15% across typical specialized workloads
- This is a one-PR change. Pattern matches your existing `multithread-mm`, `complex`, `blas` features.

**B. Per-cluster features within categories** (~30 features)
- More granular: split `nn` into `nn-conv`, `nn-pool`, `nn-norm`, `nn-activations-basic`, `nn-activations-advanced`, `nn-reduce`, `nn-softmax`, `nn-arg`; similarly for `array`/`math`/`rec`
- Saves 12–20% typical, 14% on our DFN3 case
- Bigger PR (~200 lines), slightly more cluster-boundary maintenance per new op
- Could ship as a follow-up to A

**C. Build-script introspection / proc-macro `register_for_model!("model.onnx")`**
- New crate (`tract-onnx-strip` or similar) that reads ONNX at build time and emits minimal registration
- Maximum savings, model-specific
- Substantially more code + likely community-owned to avoid burdening tract core

## Ask

Would you be interested in any of these, and if so which? We can:
1. **Submit A as a complete PR** — Cargo.toml entries + cfg annotations + a brief docs page + a `--no-default-features` CI job. ~200 lines diff, pattern matching your existing features. Ready to draft when you give the green light.
2. **Submit A + B in one PR** if you'd prefer the more granular version
3. **Stay where we are** if this isn't a direction you want tract to go

We're genuinely happy with whichever option you'd find most useful. This is downstream-WASM-consumer territory, and we know that's not the primary investment area — we'd just like to make tract a smaller dependency for the streaming-audio audience without burdening you with the upkeep, if that aligns with where you want the project to go.

If you do want help, we can have the PR for A in front of you within a week, with bench data from our DFN3 corpus showing the size delta and confirming byte-identical output.

Either way, thanks again for engaging. The +25% WASM speedup we got from your kernel work is the biggest perf win of our entire investigation — happy to give that more visibility wherever useful.

— CK (Ckristian Zoli, Vonage - part of Ericsson)

**Comment by kali** [comment]
That's worth a look, but i lack the bandwidth right now. I am a bit concerned about testability of A and/or B tbh. This could lead to a pretty big number of tract configurations in the wild that I can't see how to test. Which leads to me to think C is actually more palatable as in "this is for advanced users/deployment only, beginners will not risk stumbling on these".

Something you may want to try to reduce the footprint in browser deployment is the NNEF path. Ablating by compilation cfg flag onnx from the tract new facade is super easy (but not done at this stage).

I think C could be more relevant if we try to apply it at the nnef/tract-core level (instead of ONNX), as it make it relevant for NNEF users (and potentially other frameworks).

**Comment by czoli1976** [comment]
Thanks @kali — both points well taken.

The testability concern on A/B is fair; the combinatorial explosion is exactly why per-feature gating is brittle without a coverage matrix. Agree that C-at-the-NNEF-level is the cleaner architectural shape.

The NNEF tip is genuinely new for us — we've been ONNX-loading everything because that's what DFN3 ships, but converting at build-time and stripping the ONNX surface entirely is a much better fit for our WASM-size goal than per-op stripping. We'll try that path first and report back with measured numbers (size + RTF + bit-identity vs the ONNX-loaded build) before considering anything more ambitious.

Appreciate you taking the time on this even without bandwidth to implement.
